crates/rustc_codegen_nvvm/src/builder.rs

```diff
@@ -1134,26 +1106,126 @@ impl<'ll, 'tcx, 'a> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
     // Atomic Operations
     fn atomic_cmpxchg(
         &mut self,
-        _dst: &'ll Value,
-        _cmp: &'ll Value,
-        _src: &'ll Value,
-        _order: AtomicOrdering,
-        _failure_order: AtomicOrdering,
-        _weak: bool,
+        dst: &'ll Value,
+        cmp: &'ll Value,
+        src: &'ll Value,
+        order: AtomicOrdering,
+        failure_order: AtomicOrdering,
+        weak: bool,
     ) -> (&'ll Value, &'ll Value) {
-        // allowed but only for some things and with restrictions
-        // https://docs.nvidia.com/cuda/nvvm-ir-spec/index.html#cmpxchg-instruction
-        self.fatal("atomic cmpxchg is not supported")
+        // LLVM verifier rejects cases where the `failure_order` is stronger than `order`
+        match (order,failure_order){
+            (AtomicOrdering::SeqCst, _)=>(),
+            (_, AtomicOrdering::Relaxed)=>(),
+            (AtomicOrdering::Release, AtomicOrdering::Release) |  (AtomicOrdering::Release, AtomicOrdering::Acquire) | (AtomicOrdering::Acquire, AtomicOrdering::Acquire)=>(),
+            (AtomicOrdering::AcqRel,AtomicOrdering::Acquire)  => (),
+            (AtomicOrdering::Relaxed, _) | (_, AtomicOrdering::Release | AtomicOrdering::AcqRel | AtomicOrdering::SeqCst)=>{
+                // Invalid cmpxchg - `failure_order` is stronger than `order`! So, we abort.
+                self.abort();
+                return (self.const_undef(self.val_ty(cmp)),self.const_undef(self.type_i1()));
+            }
```

Comment:
Your match accepts (Release, Release), is that valid? The Rust atomic docs say "The (failed) load ordering can only be [SeqCst](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.SeqCst), [Acquire](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Acquire) or [Relaxed](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Relaxed)."

Reply:
It's hard for me to follow, perhaps better to list out explicit pairs and not use catch-alls?
```rust
match (order, failure_order) {
    (AtomicOrdering::Relaxed, AtomicOrdering::Relaxed)
    | (AtomicOrdering::Acquire, AtomicOrdering::Relaxed | AtomicOrdering::Acquire)
    | (AtomicOrdering::Release, AtomicOrdering::Relaxed)
    | (AtomicOrdering::AcqRel,  AtomicOrdering::Relaxed | AtomicOrdering::Acquire)
    | (AtomicOrdering::SeqCst,  AtomicOrdering::Relaxed | AtomicOrdering::Acquire | AtomicOrdering::SeqCst) => {}
    _ => { self.abort(); self.unreachable(); }
}
```

Reply:
@FractalFir Any thoughts on this? I'd like to land this as I want to build on top of it 🚀